### PR TITLE
Improve MicroProfiler and make per-op profiling the default in the benchmarks.

### PR DESCRIPTION
**Pull request by advaitjain**
* Maintain state within the MicroProfiler object such that all the logging can happen external to the interpreter.
* Refactor the benchmarks to make use of this new functionality.

This command (keyword benchmark with xtensa Fusion F1):
```
make -f tensorflow/lite/micro/tools/make/Makefile TARGET=xtensa OPTIMIZED_KERNEL_DIR=xtensa TARGET_ARCH=fusion_f1 XTENSA_CORE=F1_190305_swupgrade test_keyword_benchmark -j8
```

gives the following output at tip-of-tree:
```
InitializeKeywordRunner() took 199262 ticks (199 ms)
KeywordRunNIerations(1) took 152158 ticks (152 ms)
KeywordRunNIerations(10) took 1521087 ticks (1521 ms)
```

and with this change:
```
InitializeKeywordRunner took 152689 ticks (152 ms).

KeywordRunNIerations(1) took 152628 ticks (152 ms)
QUANTIZE took 3753 ticks (3 ms).
SVDF took 37998 ticks (37 ms).
FULLY_CONNECTED took 1363 ticks (1 ms).
SVDF took 18798 ticks (18 ms).
FULLY_CONNECTED took 1363 ticks (1 ms).
SVDF took 18798 ticks (18 ms).
FULLY_CONNECTED took 1363 ticks (1 ms).
SVDF took 18798 ticks (18 ms).
FULLY_CONNECTED took 1363 ticks (1 ms).
SVDF took 13902 ticks (13 ms).
SVDF took 15822 ticks (15 ms).
SVDF took 15822 ticks (15 ms).
FULLY_CONNECTED took 1087 ticks (1 ms).
SOFTMAX took 2037 ticks (2 ms).
QUANTIZE took 361 ticks (0 ms).

KeywordRunNIerations(10) took 1526280 ticks (1526 ms)
```

Note:
 * overall increase due to the additional book-keeping for profiling is minimal (470 ticks).
 * The initialize time has decreased because we are no longer including the time needed to compute a random input and copy it into the input buffer as part of the initialization.

Partially addresses http://b/158212576

For completeness, here is the output of the person_detection_benchmark:
```
make -f tensorflow/lite/micro/tools/make/Makefile test_person_detection_benchmark -j8
```

tip-of-tree:
```
InitializeBenchmarkRunner() took 260 ticks (0 ms)
benchmark_runner->RunSingleIteration() took 31238 ticks (31 ms)
PersonDetectionTenIerationsWithPerson() took 295380 ticks (295 ms)
PersonDetectionTenIerationsWithoutPerson() took 288194 ticks (288 ms)
```

with this change:
```
InitializeBenchmarkRunner took 302 ticks (0 ms).

WithPersonDataIterations(1) took 25590 ticks (25 ms)
DEPTHWISE_CONV_2D took 903 ticks (0 ms).
DEPTHWISE_CONV_2D took 887 ticks (0 ms).
CONV_2D took 1378 ticks (1 ms).
DEPTHWISE_CONV_2D took 453 ticks (0 ms).
CONV_2D took 1051 ticks (1 ms).
DEPTHWISE_CONV_2D took 987 ticks (0 ms).
CONV_2D took 2043 ticks (2 ms).
DEPTHWISE_CONV_2D took 225 ticks (0 ms).
CONV_2D took 1071 ticks (1 ms).
DEPTHWISE_CONV_2D took 495 ticks (0 ms).
CONV_2D took 1907 ticks (1 ms).
DEPTHWISE_CONV_2D took 118 ticks (0 ms).
CONV_2D took 938 ticks (0 ms).
DEPTHWISE_CONV_2D took 234 ticks (0 ms).
CONV_2D took 1817 ticks (1 ms).
DEPTHWISE_CONV_2D took 228 ticks (0 ms).
CONV_2D took 1909 ticks (1 ms).
DEPTHWISE_CONV_2D took 224 ticks (0 ms).
CONV_2D took 1769 ticks (1 ms).
DEPTHWISE_CONV_2D took 224 ticks (0 ms).
CONV_2D took 1797 ticks (1 ms).
DEPTHWISE_CONV_2D took 223 ticks (0 ms).
CONV_2D took 1787 ticks (1 ms).
DEPTHWISE_CONV_2D took 80 ticks (0 ms).
CONV_2D took 921 ticks (0 ms).
DEPTHWISE_CONV_2D took 99 ticks (0 ms).
CONV_2D took 1809 ticks (1 ms).
AVERAGE_POOL_2D took 8 ticks (0 ms).
CONV_2D took 2 ticks (0 ms).
RESHAPE took 1 ticks (0 ms).
SOFTMAX took 2 ticks (0 ms).

NoPersonDataIterations(1) took 25310 ticks (25 ms)
DEPTHWISE_CONV_2D took 912 ticks (0 ms).
DEPTHWISE_CONV_2D took 988 ticks (0 ms).
CONV_2D took 1356 ticks (1 ms).
DEPTHWISE_CONV_2D took 510 ticks (0 ms).
CONV_2D took 1013 ticks (1 ms).
DEPTHWISE_CONV_2D took 964 ticks (0 ms).
CONV_2D took 2004 ticks (2 ms).
DEPTHWISE_CONV_2D took 323 ticks (0 ms).
CONV_2D took 1012 ticks (1 ms).
DEPTHWISE_CONV_2D took 480 ticks (0 ms).
CONV_2D took 1770 ticks (1 ms).
DEPTHWISE_CONV_2D took 113 ticks (0 ms).
CONV_2D took 882 ticks (0 ms).
DEPTHWISE_CONV_2D took 266 ticks (0 ms).
CONV_2D took 1766 ticks (1 ms).
DEPTHWISE_CONV_2D took 219 ticks (0 ms).
CONV_2D took 1743 ticks (1 ms).
DEPTHWISE_CONV_2D took 217 ticks (0 ms).
CONV_2D took 2004 ticks (2 ms).
DEPTHWISE_CONV_2D took 217 ticks (0 ms).
CONV_2D took 1640 ticks (1 ms).
DEPTHWISE_CONV_2D took 206 ticks (0 ms).
CONV_2D took 1677 ticks (1 ms).
DEPTHWISE_CONV_2D took 52 ticks (0 ms).
CONV_2D took 1083 ticks (1 ms).
DEPTHWISE_CONV_2D took 110 ticks (0 ms).
CONV_2D took 1771 ticks (1 ms).
AVERAGE_POOL_2D took 7 ticks (0 ms).
CONV_2D took 2 ticks (0 ms).
RESHAPE took 1 ticks (0 ms).
SOFTMAX took 2 ticks (0 ms).

WithPersonDataIterations(10) took 254901 ticks (254 ms)

NoPersonDataIterations(10) took 241536 ticks (241 ms)
```

### CODE REVIEW

File: tensorflow/lite/micro/benchmarks/keyword_benchmark.cc [hunk 1]
```diff
@@ -58,32 +56,42 @@ void CreateBenchmarkRunner() {
...
+int main(int argc, char** argv) {
+  tflite::MicroProfiler profiler;
 
-TF_LITE_MICRO_BENCHMARK(InitializeKeywordRunner());
+  uint32_t event_handle = profiler.BeginEvent("InitializeKeywordRunner");
+  tflite::KeywordBenchmarkRunner* benchmark_runner =
+      CreateBenchmarkRunner(&profiler);
+  profiler.EndEvent(event_handle);
+  profiler.Log();
+  MicroPrintf("");
```
Comment (njeffrie) [inline tensorflow/lite/micro/benchmarks/keyword_benchmark.cc:87]: What are these printfs for?

File: tensorflow/lite/micro/benchmarks/person_detection_benchmark.cc [hunk 2]
```diff
@@ -43,48 +43,65 @@ alignas(16) uint8_t tensor_arena[kTensorArenaSize];
...
+
+int main(int argc, char** argv) {
+  tflite::MicroProfiler profiler;
+
+  uint32_t event_handle = profiler.BeginEvent("InitializeBenchmarkRunner");
+  tflite::PersonDetectionBenchmarkRunner* benchmark_runner =
+      CreateBenchmarkRunner(&profiler);
+  profiler.EndEvent(event_handle);
+  profiler.Log();
+  MicroPrintf("");
```
Comment (njeffrie) [inline tensorflow/lite/micro/benchmarks/person_detection_benchmark.cc:84]: Same question about these printfs. Are they necessary for flushing or something? Perhaps a comment if they are necessary.

**Review by njeffrie — APPROVED** [review]

File: tensorflow/lite/micro/benchmarks/keyword_benchmark.cc [hunk 3]
```diff
@@ -58,32 +56,42 @@ void CreateBenchmarkRunner() {
...
-TF_LITE_MICRO_BENCHMARKS_BEGIN
+int main(int argc, char** argv) {
+  tflite::MicroProfiler profiler;
 
-TF_LITE_MICRO_BENCHMARK(InitializeKeywordRunner());
+  uint32_t event_handle = profiler.BeginEvent("InitializeKeywordRunner");
+  tflite::KeywordBenchmarkRunner* benchmark_runner =
+      CreateBenchmarkRunner(&profiler);
+  profiler.EndEvent(event_handle);
+  profiler.Log();
```
[the remark this answers is not in the document]
Comment (advaitjain) [reply tensorflow/lite/micro/benchmarks/keyword_benchmark.cc:86]: we need this profile.Log to los the Begin/EndEvent from lines 82-85.

File: tensorflow/lite/micro/benchmarks/keyword_benchmark.cc (hunk 1 above)
Comment (advaitjain) [reply tensorflow/lite/micro/benchmarks/keyword_benchmark.cc:87]: added a comment

File: tensorflow/lite/micro/benchmarks/person_detection_benchmark.cc (hunk 2 above)
Comment (advaitjain) [reply tensorflow/lite/micro/benchmarks/person_detection_benchmark.cc:84]: added a comment.

File: tensorflow/lite/micro/micro_profiler.h [hunk 4]
```diff
@@ -16,55 +16,98 @@ limitations under the License.
...
 namespace tflite {
 
 // MicroProfiler creates a common way to gain fine-grained insight into runtime
 // performance. Bottleck operators can be identified along with slow code
 // sections. This can be used in conjunction with running the relevant micro
 // benchmark to evaluate end-to-end performance.
+class MicroProfiler {
+ public:
+  MicroProfiler() = default;
+  ~MicroProfiler() = default;
```
[the remark this answers is not in the document]
Comment (advaitjain) [reply tensorflow/lite/micro/micro_profiler.h:30]: done.

**Pull request merged**